### ISSUE TITLE
Issue with SuRandomAccessFile and SuFile

### ISSUE THREAD

**Issue by PerfectSlayer** [topjohnwu/libsu#58]
Hello John,

I tried few things with `SuRandomAccessFile` and I find an issue with `SuFile`.

Here is the code I tried which works fine with a `File`:

```java
        try {
            SuRandomAccessFile file = SuRandomAccessFile.open(new File("/system/etc/hosts"), "r");
            String line;
            while ((line = file.readLine()) != null) {
                Log.e("TEST", line);
            }
        } catch (IOException e) {
            Log.e("TEST", "error", e);
        }
```

But if I use a `SuFile` instead of a java `File`:

```java
        try {
            SuRandomAccessFile file = SuRandomAccessFile.open(new SuFile("/system/etc/hosts"), "r");
            String line;
            while ((line = file.readLine()) != null) {
                Log.e("TEST", line);
            }
        } catch (IOException e) {
            Log.e("TEST", "error", e);
        }
```

I always got an error:

```
2020-08-16 11:44:24.736 21996-21996/com.topjohnwu.libsuexample D/LIBSU: java.io.InterruptedIOException
        at com.topjohnwu.superuser.internal.ShellImpl$DefaultTask.run(ShellImpl.java:280)
        at com.topjohnwu.superuser.internal.ShellImpl.execTask(ShellImpl.java:233)
        at com.topjohnwu.superuser.internal.JobImpl.exec0(JobImpl.java:53)
        at com.topjohnwu.superuser.internal.JobImpl.exec(JobImpl.java:70)
        at com.topjohnwu.libsuexample.MainActivity$ExampleInitializer.onInit(MainActivity.java:84)
        at com.topjohnwu.superuser.internal.BuilderImpl.build(BuilderImpl.java:84)
        at com.topjohnwu.superuser.internal.BuilderImpl.build(BuilderImpl.java:58)
        at com.topjohnwu.superuser.internal.MGR.getShell(MGR.java:41)
        at com.topjohnwu.superuser.Shell.getShell(Shell.java:148)
        at com.topjohnwu.superuser.ShellUtils.fastCmdResult(ShellUtils.java:85)
        at com.topjohnwu.superuser.io.SuFile.cmdBool(SuFile.java:110)
        at com.topjohnwu.superuser.io.SuFile.exists(SuFile.java:184)
        at com.topjohnwu.superuser.internal.ShellIO.<init>(ShellIO.java:57)
        at com.topjohnwu.superuser.internal.ShellIO.get(ShellIO.java:75)
        at com.topjohnwu.superuser.internal.IOFactory.createShellIO(IOFactory.java:41)
        at com.topjohnwu.superuser.io.SuRandomAccessFile.open(SuRandomAccessFile.java:54)
        at com.topjohnwu.libsuexample.MainActivity.runIoTests(MainActivity.java:244)
        at com.topjohnwu.libsuexample.MainActivity.lambda$onCreate$5$MainActivity(MainActivity.java:219)
        at com.topjohnwu.libsuexample.-$$Lambda$MainActivity$a7zsI796bNp0TD7tk4yhXst0ak4.onClick(Unknown Source:2)
        at android.view.View.performClick(View.java:7259)
        at android.view.View.performClickInternal(View.java:7236)
        at android.view.View.access$3600(View.java:801)
        at android.view.View$PerformClick.run(View.java:27892)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
     Caused by: java.util.concurrent.ExecutionException: java.lang.NullPointerException: Attempt to invoke virtual method 'int java.lang.String.length()' on a null object reference
        at java.util.concurrent.FutureTask.report(FutureTask.java:123)
        at java.util.concurrent.FutureTask.get(FutureTask.java:193)
        at com.topjohnwu.superuser.internal.ShellImpl$DefaultTask.run(ShellImpl.java:277)
        at com.topjohnwu.superuser.internal.ShellImpl.execTask(ShellImpl.java:233) 
        at com.topjohnwu.superuser.internal.JobImpl.exec0(JobImpl.java:53) 
        at com.topjohnwu.superuser.internal.JobImpl.exec(JobImpl.java:70) 
        at com.topjohnwu.libsuexample.MainActivity$ExampleInitializer.onInit(MainActivity.java:84) 
        at com.topjohnwu.superuser.internal.BuilderImpl.build(BuilderImpl.java:84) 
        at com.topjohnwu.superuser.internal.BuilderImpl.build(BuilderImpl.java:58) 
        at com.topjohnwu.superuser.internal.MGR.getShell(MGR.java:41) 
        at com.topjohnwu.superuser.Shell.getShell(Shell.java:148) 
        at com.topjohnwu.superuser.ShellUtils.fastCmdResult(ShellUtils.java:85) 
        at com.topjohnwu.superuser.io.SuFile.cmdBool(SuFile.java:110) 
        at com.topjohnwu.superuser.io.SuFile.exists(SuFile.java:184) 
        at com.topjohnwu.superuser.internal.ShellIO.<init>(ShellIO.java:57) 
        at com.topjohnwu.superuser.internal.ShellIO.get(ShellIO.java:75) 
        at com.topjohnwu.superuser.internal.IOFactory.createShellIO(IOFactory.java:41) 
        at com.topjohnwu.superuser.io.SuRandomAccessFile.open(SuRandomAccessFile.java:54) 
        at com.topjohnwu.libsuexample.MainActivity.runIoTests(MainActivity.java:244) 
        at com.topjohnwu.libsuexample.MainActivity.lambda$onCreate$5$MainActivity(MainActivity.java:219) 
        at com.topjohnwu.libsuexample.-$$Lambda$MainActivity$a7zsI796bNp0TD7tk4yhXst0ak4.onClick(Unknown Source:2) 
        at android.view.View.performClick(View.java:7259) 
        at android.view.View.performClickInternal(View.java:7236) 
        at android.view.View.access$3600(View.java:801) 
        at android.view.View$PerformClick.run(View.java:27892) 
        at android.os.Handler.handleCallback(Handler.java:883) 
        at android.os.Handler.dispatchMessage(Handler.java:100) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.app.ActivityThread.main(ActivityThread.java:7356) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930) 
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'int java.lang.String.length()' on a null object reference
        at com.topjohnwu.superuser.internal.StreamGobbler.isEOS(StreamGobbler.java:46)
        at com.topjohnwu.superuser.internal.StreamGobbler$OUT.call(StreamGobbler.java:77)
        at com.topjohnwu.superuser.internal.StreamGobbler$OUT.call(StreamGobbler.java:66)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
```

It comes from the `StreamGobbler` which tries to find the EOS when the `BufferedReader` reaches EOF first.
I think it does not find the `eos` tag and `isEOS` fails.

Moreover, there is something strange with the `isEOS` implementation:

```java
protected boolean isEOS(String line) {
        boolean eof = false;
        int sl = line.length() - 1;
        int tl = eos.length() - 1;
        if (sl >= tl) {
            eof = true;
            for (; tl >= 0; --tl, --sl) {
                if (eos.charAt(tl) != line.charAt(sl)) {
                    eof = false;
                    break;
                }
            }
            if (eof)
                line = sl >= 0 ? line.substring(0, sl + 1) : null;
        }
        if (list != null && line != null) {
            list.add(line);
            Utils.log(TAG, line);
        }
        return eof;
    }
```
`line` is never tested to be `null` at method start where it can becomes `null` and nullability is checked later.

Long story short: I might be using `SuFile` the wrong way. If not, there might be a bug with `SuRandomAccessFile` 

Regards,
Bruce

**Comment by topjohnwu** [comment]
There is indeed a bug in `ShellIO` that causes it to read incorrect data for readline (never tested that code so it got ignored from me), however I do not see the issue you experienced (cannot reproduce).
From `StreamGobbler`'s code, it is literally impossible to have it return null unless the shell itself is dead.
It might be a good idea to check null so at least it won't crash, but to be honest this should really never happen.
May you elaborate on more details?

**Comment by topjohnwu** [comment]
@PerfectSlayer `SuRandomAccessFile` is fixed, but as your problem does not lie in the I/O layer, do you have more details?
Basically, your shell is running into a state that should never happen.

**Comment by PerfectSlayer** [comment]
I just push a branch _issue-sufile_ https://github.com/PerfectSlayer/libsu/commits/issue-sufile into my fork with a modified sample app.
I update the two first buttons to start two tests:
- The first using java.io File: working
- The second using SuFile: failing

![image](https://user-images.githubusercontent.com/1766222/90436186-8658b000-e0d0-11ea-8ca0-d949e77b6b07.png)

You should be able to reproduce with it. Let me know if you still can't trigger the NPE.

**Comment by MuntashirAkon** [comment]
> I seem the `com.topjohnwu.libsuexample.MainActivity.ExampleInitializer` load a script and it fails to load/parse.

`Shell.Initializer` seems to throw a `InterruptedIOException` if it contains a script that has outputs.

**Comment by topjohnwu** [comment]
@PerfectSlayer I am 99.999% sure this is not something that `libsu` can fix, rather there are some incompatibilities with the shell script and your shell (for some weird reason). 

I have a very plausible theory: are you using a Windows machine that checks out the shell file in CRLF rather than LF?

The exact same setup is used in the Magisk app, and if that app is working fine, then there is no issue in the code itself.

**Comment by topjohnwu** [comment]
@PerfectSlayer this shall prevent it from happening again... https://github.com/topjohnwu/libsu/commit/713e52c81637da8bd2f1460122c5e4688dd58bae

**Comment by PerfectSlayer** [comment]
Thanks for the fix, it should do the trick! 👍 
There is another issue with you build script when building on Windows, I will make an issue to address it.
